woocommerce/woocommerce-gateway-stripe
# Fix: PaymentMethods of type card cannot be attached to Customers directly without 3DS due to Indian payment regulations

Fixes #2664 

Subscription purchases and recurring payments are not working for INR. The transaction fails with the following error-

> invalid_request_error
PaymentMethods of type card cannot be attached to Customers directly without 3DS due to Indian payment regulations. Please instead provide the PaymentMethod and Customer alongside a SetupIntent or PaymentIntent with the setup_future_usage parameter. See https://support.stripe.com/questions/guide-for-saving-cards-in-india for more details.

## Changes proposed in this Pull Request:
- The issue occurs for the attach customer request. Removed the code that calls the `attach_source` function (which then makes the API call to attach the customer) from the checkout flow (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2964/commits/1d0ed3b53fc243a0d02b56afd18fb3b6ffe9f947) and add payment method from `My Account` page flow (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2964/commits/8542f66afd0ebd4cf4b6747abac4a31984e0af40)
- Removing the request does not have any impact and does not break any flow. As in both of the flows we create an intent down the line with `setup_future_usage` which on the Stripe end attaches the payment method to the customer (even without any request sent from our end)

## Testing instructions
### How to reproduce
- Disable the `New checkout experience` (UPE)
- Connect to an Indian Stripe account. (the error can be reproduced only for the Indian Stripe accounts)
- Set store currency to INR.
- As a shopper add a subscription product to the cart and go to the checkout page.
- Try to purchase the subscription with an [India issued card](https://docs.stripe.com/india-recurring-payments?integration=paymentIntents-setupIntents#testing) (`4000003560000123`)
- You should encounter an error and payment should fail.
- If you check the logs in your Stripe dashboard you should find an error for the attach payment method request (`POST /v1/payment_methods/pm_1OwHVQSGrDfhP3zSiJ5xVeAC/attach`)

> invalid_request_error
PaymentMethods of type card cannot be attached to Customers directly without 3DS due to Indian payment regulations. Please instead provide the PaymentMethod and Customer alongside a SetupIntent or PaymentIntent with the setup_future_usage parameter. See https://support.stripe.com/questions/guide-for-saving-cards-in-india for more details.


### Test the fix
1. Disable the `New checkout experience` (UPE)
2. Connect to an Indian Stripe account.
3. Set store currency to INR.
4. As a shopper add a subscription product to the cart and go to the checkout page.
5. Try to purchase the subscription with an [India issued card](https://docs.stripe.com/india-recurring-payments?integration=paymentIntents-setupIntents#testing) (`4000003560000123`)
6. The payment should be successful.
7. Find the `_stripe_mandate_id` order meta for this order in your database and ensure that it has a value.
8. In your Stripe dashboard, find this customer and ensure that the card is attached to the customer.
9. Now enable the `New checkout experience` (UPE). Test steps 2-8 and ensure that there is no error.

### Test saving card
- Test saving a regular card (`4242424242424242`) and an India issued card (`4000003560000123`) from `My Account` page for both UPE enabled and disabled state.

### Test creating an account
1. As a new shopper add a simple product to cart. 
2. On the checkout page select create new account and save payment method checkbox.
3. Complete the purchase with an India issued card (`4000003560000123`).
4. In your Stripe dashboard, find this customer and ensure that the card is attached to the customer.
5. Test steps 1-4 with a regular card (`4242424242424242`)
6. Test steps 1-5 for both UPE enabled and disabled state. 

### Additional testing
- Test any other flow which needs attaching a payment to a customer or saving a payment method that you can think of.
- Test the flows for a non Indian Stripe account as well.
